fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/MessageValue.java

```diff
@@ -67,14 +70,32 @@ public static Object getFieldValue(@Nonnull MessageOrBuilder message, @Nonnull L
      * Get the value of the field with the given field name on the given message.
      * If the field is repeated, the repeated values are combined into a list. If the field has a message type,
      * the value is returned as a {@link Message} of that type. Otherwise, the field is returned as a primitive.
+     *
      * @param message a message or builder to extract the field from
      * @param fieldName the field to extract
+     *
      * @return the value of the field as described above
      */
     @Nullable
     public static Object getFieldOnMessage(@Nonnull MessageOrBuilder message, @Nonnull String fieldName) {
         final Descriptors.FieldDescriptor field = findFieldDescriptorOnMessage(message, fieldName);
+        return getFieldOnMessage(message, field);
+    }
+
+    @Nullable
+    public static Object getFieldOnMessage(@Nonnull MessageOrBuilder message, int fieldNumber) {
+        final Descriptors.FieldDescriptor field = findFieldDescriptorOnMessage(message, fieldNumber);
+        return getFieldOnMessage(message, field);
+    }
+
+    @Nullable
+    private static Object getFieldOnMessage(@Nonnull MessageOrBuilder message, @Nonnull Descriptors.FieldDescriptor field) {
         if (field.isRepeated()) {
+            if (!message.getAllFields().containsKey(field)) {
+                // if a repeated field is unset, should return null
+                return null;
+            }
```

Comment:
⚠️ This line is seems to change record layer behaviour; Previously and empty repeated field is considered `equal` to an empty list, now it returns null.

(the parametrised test `com.apple.foundationdb.record.query.expressions.QueryExpressionTest#testParameterComparisons` is now red, since this particular test is now failing because of this:

```java
{ "{} EQUALS {}", "repeat_me", Collections.emptyList(), Comparisons.Type.EQUALS, Collections.emptyList(), Boolean.TRUE},
```

I recommend checking with more experienced colleagues whether this is ok.

Reply:
@pengpeng-lu what was the conclusion about this particular change during our nullability discussion?

Reply:
It was incorrect -> after this change, the field is null if a field is unset or set to empty array. Normen and I discussed about the Type system, and I just sent what I think should be done in our slack.